src/server/hset_family.cc

```diff
@@ -1020,25 +1037,86 @@ void HSetFamily::HRandField(CmdArgList args, ConnectionContext* cntx) {
     StringVec str_vec;
 
     if (pv.Encoding() == kEncodingStrMap2) {
-      // TODO: to create real random logic.
       StringMap* string_map = (StringMap*)pv.RObjPtr();
 
-      sds key = string_map->begin()->first;
-      str_vec.emplace_back(key, sdslen(key));
+      // Randomize list
+      std::vector<dfly::StringMap::iterator> it_vec;
+      for (auto it = string_map->begin(); it != string_map->end(); ++it) {
+        it_vec.push_back(it);
+      }
+      auto rng = std::default_random_engine{};
+      std::shuffle(std::begin(it_vec), std::end(it_vec), rng);
+
+      if (count > 0) {
+        // If positive, emplace only unique elements, capped by vector size
+        for (size_t i = 0; i < std::min(size_t(count), it_vec.size()); i++) {
+          auto it = it_vec[i];
+          sds it_key = it->first;
+          sds it_value = it->second;
+          str_vec.emplace_back(it_key, sdslen(it_key));
+          if (with_values) {
+            str_vec.emplace_back(it_value, sdslen(it_value));
+          }
+        }
+      } else if (count < 0) {
+        // If negative, randomly pick elements |COUNT| times
+        for (size_t i = 0; i < size_t(-count); i++) {
+          auto it = it_vec[rand() % it_vec.size()];
+          sds it_key = it->first;
+          sds it_value = it->second;
+          str_vec.emplace_back(it_key, sdslen(it_key));
+          if (with_values) {
+            str_vec.emplace_back(it_value, sdslen(it_value));
+          }
+        }
+      } else if (it_vec.size() > 0) {
+        // Special case if count is 0, just single element in the array
+        sds it_key = it_vec[0]->first;
+        sds it_value = it_vec[0]->second;
+        str_vec.emplace_back(it_key, sdslen(it_key));
+        if (with_values) {
+          str_vec.emplace_back(it_value, sdslen(it_value));
+        }
+      }
```

Comment:
Your 3 blocks look suspiciously similar 🙂 🔎 

They can be replaced with a something like this
```c++
for (size_t i = 0; i < std::min(size_t(abs(count)), it_vec.size()); i++) {
  size_t idx = rng() % it_vec.size();
  auto [it_key, it_value] = it_vec[idx];

  str_vec.emplace_back(it_key);
  if (with_values)
    str_vec.emplace(it_value);

  if (count > 0) { // swap & pop back
    swap(it_vec[idx], it_vec.back());
    it_vec.pop_back();
  }
}
```

Second, if we always choose only the first element for case 3, then lets not iterate over the whole set for it.

Reply:
Also note that you always traverse the whole map, even with count=2. Given that the feature was requested and listpack is actually not faster (because it needs to seek), it's fine in my opinion. The more optimal version requires digging into StringMap to add it there 😵